cmd/repo-brancher/main.go

```diff
@@ -0,0 +1,167 @@
+		logger := config.LoggerForInfo(*repoInfo)
+		if (o.Org != "" && o.Org != repoInfo.Org) || (o.Repo != "" && o.Repo != repoInfo.Repo) {
+			return nil
+		}
+		if !(promotion.PromotesOfficialImages(configuration) && configuration.PromotionConfiguration.Name == o.CurrentRelease) {
+			return nil
+		}
+
+		repoDir := path.Join(gitDir, repoInfo.Org, repoInfo.Repo)
+		if err := os.MkdirAll(repoDir, 0775); err != nil {
+			logger.WithError(err).Fatal("could not ensure git dir existed")
+			return nil
+		}
+
+		futureBranchForCurrentPromotion, futureBranchForFuturePromotion, err := promotion.DetermineReleaseBranches(o.CurrentRelease, o.FutureRelease, repoInfo.Branch)
+		if err != nil {
+			logger.WithError(err).Error("could not determine future branch that would promote to current imagestream")
+			return nil
+		}
+
+		remote, err := url.Parse(fmt.Sprintf("https://github.com/%s/%s", repoInfo.Org, repoInfo.Repo))
+		if err != nil {
+			logger.WithError(err).Fatal("Could not construct remote URL.")
+		}
+		if o.Confirm {
+			remote.User = url.UserPassword(o.username, o.password)
+		}
+		for _, command := range [][]string{{"init"}, {"fetch", "--depth", "10", remote.String(), repoInfo.Branch}} {
+			cmdLogger := logger.WithFields(logrus.Fields{"commands": fmt.Sprintf("git %s", strings.Join(command, " "))})
+			cmd := exec.Command("git", command...)
+			cmd.Dir = repoDir
+			cmdLogger.Debug("Running command.")
+			if out, err := cmd.CombinedOutput(); err != nil {
+				cmdLogger.WithError(err).WithFields(logrus.Fields{"output": string(out)}).Error("Failed to execute command.")
+				return nil
+			} else {
+				cmdLogger.WithFields(logrus.Fields{"output": string(out)}).Debug("Executed command.")
+			}
+		}
+
+		// when we're initializing the branch, we just want to make sure
+		// it is in sync with the current branch that is promoting
+		for _, futureBranch := range []string{futureBranchForCurrentPromotion, futureBranchForFuturePromotion} {
+			if futureBranch == repoInfo.Branch {
+				continue
+			}
+			branchLogger := logger.WithField("future-branch", futureBranch)
+			command := []string{"ls-remote", remote.String(), fmt.Sprintf("refs/heads/%s", futureBranch)}
+			cmdLogger := branchLogger.WithFields(logrus.Fields{"commands": fmt.Sprintf("git %s", strings.Join(command, " "))})
+			cmd := exec.Command("git", command...)
+			cmd.Dir = repoDir
+			cmdLogger.Debug("Running command.")
+			if out, err := cmd.CombinedOutput(); err != nil {
+				cmdLogger.WithError(err).WithFields(logrus.Fields{"output": string(out)}).Error("Failed to execute command.")
+				continue
+			} else {
+				cmdLogger.WithFields(logrus.Fields{"output": string(out)}).Debug("Executed command.")
+				if string(out) == "" && !o.fastForward {
+					branchLogger.Info("Remote already has branch, skipping.")
+					continue
+				}
+			}
+
+			if !o.Confirm {
+				branchLogger.Info("Would create new branch.")
```

Comment:
The message was a little confusing for me when I first tried to run the tool with `--fast-forward` but without `--confirm`, and I knew the branch should exist already.